src/XR/features/WebXRControllerMovement.ts

```diff
@@ -0,0 +1,547 @@
+import { WebXRAbstractFeature } from "./WebXRAbstractFeature";
+import { MotionControllerComponentType } from "../motionController/webXRAbstractMotionController";
+import { Tools } from "../../Misc/tools";
+
+/**
+ * The options container for the controller movement module
+ */
+export interface IWebXRControllerMovementOptions {
+    /**
+     * Override default behaviour and provide your own movement controls
+     */
+    customRegistrationConfigurations?: WebXRControllerMovementRegistrationConfiguration[];
+    /**
+     * Is movement enabled
+     */
+    movementEnabled?: boolean;
+    /**
+    * Camera direction follows view pose and movement by default will move independently of the viewer's pose.
+    */
+    movementOrientationFollowsViewerPose: boolean;
+    /**
+     * Movement speed factor (default is 1.0)
+     */
+    movementSpeed?: number;
+    /**
+    * Minimum threshold the controller's thumbstick/touchpad must pass before being recognized for movement (avoids jitter/unintentional movement)
+    */
+    movementThreshold?: number;
+    /**
+     * Is rotation enabled
+     */
+    rotationEnabled?: boolean;
+    /**
+    * Minimum threshold the controller's thumstick/touchpad must pass before being recognized for rotation (avoids jitter/unintentional rotation)
+    */
+    rotationThreshold?: number;
+    /**
+     * Movement speed factor (default is 1.0)
+     */
+    rotationSpeed?: number;
+    /**
+     * Babylon XR Input class for controller
+     */
+    xrInput: WebXRInput;
+}
+
+/**
+ * Feature context is used in handlers and on each XR frame to control the camera movement/direction.
+ */
+export type WebXRControllerMovementFeatureContext = {
+    movementEnabled: boolean
+    movementOrientationFollowsViewerPose: boolean
+    movementSpeed: number
+    movementThreshold: number
+    rotationEnabled: boolean
+    rotationSpeed: number
+    rotationThreshold: number
+};
+
+/**
+ * Current state of Movements shared across components and handlers.
+ */
+export type WebXRControllerMovementState = {
+    moveX: number
+    moveY: number
+    rotateX: number
+    rotateY: number
+};
+
+/**
+ * Button of Axis Handler must be specified.
+ */
+export type WebXRControllerMovementRegistrationConfiguration = {
+    /**
+     * handlers are filtered to these types only
+     */
+    allowedComponentTypes?: MotionControllerComponentType[]
+    /**
+     * For registering movement to specific hand only.  Useful if your app has a "main hand" and "off hand" for determining the functionality of a controller.
+     */
+    handedness?: XRHandedness
+    /**
+     * For main component only (useful for buttons and may not trigger axis changes)
+     */
+    mainComponentOnly?: boolean
+    /**
+     * Additional predicate to apply to controllers to restrict a handler being added.
+     */
+    componentSelectionPredicate?: (xrController: WebXRInputSource) => Nullable<WebXRControllerComponent>
+} & ({

```

Comment:
formatting feels a bit off here.

Reply:
I ran it through the Prettier Playground!!  I saw you added a .prettierrc some time ago and then it doesn't look like it's still there.  tslint was deprecated in 2019.  would be interesting eslint + prettier, but even on small projects that is a massive undertaking.